### ISSUE TITLE
Spec for nested fields

### ISSUE THREAD

**Issue by eric-burel** [VulcanJS/Vulcan#2416]
**The issue:** the notion of nested object is ambiguous, GraphQL and SimpleSchema have different intepretations.

In GraphQL you can either be a blackbox JSON or a nested type, there is no intermdiate. But by JSON, they don't mean a JS object, they mean an actual JSON object. So JavaScript native objects like `Date` within a `JSON` will be serialized by a call to `JSON.stringify` before being sent to the server, but won't be automatically deserialized by the `JSON.parse` server side. This will fail SimpleSchema validation which excepts and actual `Date` and not a date isostring.

See https://github.com/VulcanJS/Vulcan/issues/2211

The alternative is to create an explicit GraphQL type for your object. Thanks to scalar, GraphQL is able to deserialize the object correctly on the server. But that does not apply to JSON, only to custom type.

So we needed to implement schema nesting to provide more alternative to the user to remove the ambiguity of nested schemas and object. This is less magic but also less bugs.

This is the new proposed possible behaviours, depending on the field schema:

1. **Preferred**  (but breaking change in custom fragments): `address: { type: new SimpleSchema(...) }`: will generate a nested graphQL type and default fragments, tied to the collection (say `CustomerAdress` for the `customer` collection). This is a breaking change if you use custom fragments, you'll need to update them. The warnings are rather explicit so migration is easy. You will have a clean schema.
2. **Alternative** (no breaking change): `address: { type: new SimpleSchema(...), typeName: "JSON" }`: Will behave the same as the API did previously, but it's more explicit. Now the user is aware that the type is JSON, so Date validation won't work. This is the second migration option, closer to what exists now (with the same limitations too).
3. **Intl** (works the same as before): `address: { type: String, intl: true) }`:  Intl strings are a edge case with their own behaviour (they will generate intermediate fields).
4. **Actual JSON** (for some edge cases, works the same as before): `address: { type: Object }`: a JSON blackbox. No nested schema, no validation. Eg if you have huge object or something very technical.

You can also blackbox your schema:

5. **Not advised** (does not make sense to provide a schema here): `address: { type: new SimpleSchema(...), blackbox: true }`: No nested schema, no validation. Does not make much sense as both nested graphql schema generation and simpleSchema validation are disabled.
6. **Not advised and not officially supported** (does not make sense to provide a schema here): `address: { type: new SimpleSchema(...), typeName: "FooBar" }`: won't generate a nested graphQL type. Expects FooBar to exists. `addGraphQLTypeAndResolver` will help creating the `FooBar` type. This is not public yet and experimental. It currently lacks validation, which in turns need to register the schema somewhere. Apollinaire started to experiment this, the advantage is that you will be able to reuse a nested schema accross collections (so `Address` instead of `CustomerAddress`, `FactoryAddress`....

And in the future we will have:

7. **Reuse a nested schema** (NOT YET IMPLEMENTED): `{address: { typeName: "FooBar" }` This will load the schema associated to the "FooBar" type. This require registering the schema somewhere. This will allow to reuse a graphQL type tied to a schema, without tying it to an actual collection. Great for addresses for example.

Note: this apply to objects within arrays too (`addresses: { type: Array}, "addresses.$": { type: ... }`)

Edit: updated the spec to clarify
Edit: the questions we need to solve:

- The `typeName` syntax is confusing, especially about how it interacts with `type`. What could be a better naming? Note that `resolveAs` already uses the `typeName` option for the graphql type
- Should we tie the `typeName` to the schema?
- Should we reuse the `resolveAs: { typeName }` option, which already exists, instead of introducing a new `typeName` field?

**Comment by SachaG** [comment]
Great summary. Some random thoughts: 

- Is it weird to have both `type` and `typeName`? Especially since with this syntax we can now have `typeName` both on a field and on a field's `resolveAs`. I feel like that might get confusing?
- Not sure I understand the different between scenarios 2, 4, and 5. Maybe we should clearly emphasize which one we think is best (provided people don't want the nested GraphQL schema generation). 

To avoid my first concern I would suggest this alternate pattern:

```js
const FooSchema = new SimpleSchema(...)

FooSchema.typeName = "FooBar"

// ...

address: { type: FooSchema }
```

This has the added benefit that schema `FooSchema` will have the same type wherever it's used, which feels more consistent than leaving it up to be specified in each different field.

**Comment by SachaG** [comment]
In that case should it be more like this? 

```
address: { 
  type: new SimpleSchema(...), 
  resolveAs: {
    typeName: "FooBar" 
  }
}
```

Or is that even more confusing?

**Comment by eric-burel** [comment]
Also the `typeName` may not be tied to a nested field, this should actually be usable by other fields to  bypass the default behaviour. For example if you want the GraphQL type to be a new kind of custom Scalar. So I am not sure we would be able to tie it to the nested schema, it's more something related to the field itself.

**Comment by kevinashworth** [comment]
I am using this, which seems a lot like **1. Preferred** -- but, if so, it is not a breaking change. I must be missing something?

```
const addressSchema = {
  street1: {
    type: String,
    ...
  },
  city: {
    type: String,
    .
  },
  zip: {
    type: String,
    ...
  }
}

export const addressSubSchema = new SimpleSchema(addressSchema)
```

...

```  
import { addressSubSchema } from '../shared_schemas.js'
...
addresses: {
    type: Array,
    ... },
  'addresses.$': {
    type: addressSubSchema
  }
```

**Comment by eric-burel** [comment]
@kevinashworth Yep but this version won't generate the GraphQL subschema. Internally this will be a blackbox `JSON`. This is fine, excepts if you start adding native types in your nested schema, like a `Date`. If you add a Date field in `addressSubSchema` you will probably encounter this: https://github.com/VulcanJS/Vulcan/issues/2211

So solution 1 will actually generate the graphQL schema, solves #2211, but will require you to improve your custom fragments (if you don't use custom fragments you'll be fine). 
Solution 2 will behave the same as previously, but without magic, so that you are aware that your graphQL type is `JSON` for `addresses.$`, with all the limitations it can implies like not supporting Dates correctly.

**Comment by kevinashworth** [comment]
Cool. I think I get it now. And if I understand now, I vote for **1.**

**Comment by eric-burel** [comment]
Thanks! Other alternative will also be working actually, they can make sense if you have some very specific use case. 

Especially the number 7. (the one that is not yet implemented) will allow to reuse a nested graphql type accross collections

**Comment by SachaG** [comment]
> For example if you want the GraphQL type to be a new kind of custom Scalar.

That's true, but that's what `resolveAs` is for currently. That's been how we've maintained the distinction between the "database" field and the "graphQL" field. I think that's why it feels weird to me to now have `typeName` outside of `resolveAs`, it blurs the line between the two. I think everything that affects the GraphQL API one way or another (and only affects that) should stay inside `resolveAs` to be consistent.

**Comment by eric-burel** [comment]
Right, I'll checkout if I can update the branch with this behaviour. However how to handle arrays of resolveAs? I have trouble figuring when  `resolveAs.typeName` should become the field type (if we need addOriginalField, or maybe if the resolveAs has no resolver?), that's the only thing that bothers me with this approach. I've mentally associated resolveAs with "I need one or more custom field resolver" instead of the graphql part of the field schema.

**Comment by SachaG** [comment]
Yeah it's true that resolveAs arrays make everything a bit more complicated… maybe it wasn't a good idea to add that feature after all? In any case for now I think we can just pick the first `typeName` or something, we can just tell people not to combine custom sub-schemas and resolveAs arrays together for now. And once we've used this a bit more, we can introduce a better API.

**Comment by eric-burel** [comment]
We may even keep this `typeName` option "secret" until we figure out the best syntax. Right now it's basically an escape hatch not to update custom fragments. Using "JSON" type for an actual nested object is not a good practice anyway. 

So maybe you can announce the preferred option 1. as a breaking change (but honestly one of the easiest to solve, you just have to add fields to your nested objects as you would do in any GraphQL app, the warnings are explicits), and we just tell people with big app that they can use `typeName` to bypass this, but that it may be broken in next releases?

**Comment by eric-burel** [comment]
The `resolveAs` array is a rather smart way to load multiple data from a single field. The alternative would be multiple fields that use the "document" parameter in their "resolver" ?

**Comment by SachaG** [comment]
> We may even keep this `typeName` option "secret" until we figure out the best syntax. Right now it's basically an escape hatch not to update custom fragments. Using "JSON" type for an actual nested object is not a good practice anyway.
> 
> So maybe you can announce the preferred option 1. as a breaking change (but honestly one of the easiest to solve, you just have to add fields to your nested objects as you would do in any GraphQL app, the warnings are explicits), and we just tell people with big app that they can use `typeName` to bypass this, but that it may be broken in next releases?

OK



> The `resolveAs` array is a rather smart way to load multiple data from a single field. The alternative would be multiple fields that use the "document" parameter in their "resolver" ?

Yeah, it was getting a bit verbose when for example you have a `createdAt` date that you want to format in different styles (`createdAtFormattedShort`, `createdAtFormattedFull`, etc.) so that gave me the idea for adding support for arrays of resolved fields. There is no strict relation between the original field and the resolved field (the resolve field can return any data including the value of a completely different field) so it's more about keeping your code clean.
